### ISSUE TITLE
PDF Import: automatically figure out the date of the document

### ISSUE THREAD

**Issue by jflesch** [openpaperwork/paperwork#338]
PDF have metadatas, including "creation date" and "last modified". These dates could be used for the document date.

Note: be careful, some PDFs have crappy dates

**Comment by Whisprin** [comment]
Maybe paperwork should actually keep two dates:
- scanned/imported at date
- document date

One could even separate the scanned and imported at date. I am scanning documents as they come and import them in bulks into paperwork. The imported date is even less interesting to me than the scanned date (e.g. creation date of the pdf).

Also the document date shouldn't be set automatically. Paperwork should suggest dates based on file contents.
1. Do OCR
2. Return all strings matching ([0-9]?[0-9]).([0-9]?[0-9]).(20)?[0-9][0-9]
3. Suggest a list of documents dates in the date chooser
4. Optional: Set document date to best guess (first hit, date closest to current date, etc.)
5. The regex could be user configurable

If there's a better way to talk about features please let me know.

**Comment by jflesch** [comment]
Hm, below the calendar we could display a list of suggested dates based on the content of the document. I guess it could help first time users quite a lot.

One of the main problems will be to find the dates in the document. Each culture has its own way of writing the dates ... :/
The regex you provided will catch all the dates written numerically in occidental cultures I think, but not the ones with words ("Septembre the 5th 2012" or "5 septembre 2012").

It's not an easy problem, but it would be a fun one to try to solve :-)

**Comment by Whisprin** [comment]
Yes, you're right. Didn't think about different conventions. Even the order of month and day isn't fixed.

But parsing a date, without knowing the format already, from a human readable string sounds like something others have already taken on.

So there are two problems:
- Find a string which could be a date
- Parse that string into a date, taking locale and user preferences into account

Looking at my local paperwork db: I could try to extract dates from a paper.words file on the command line.

**Comment by tYYGH** [comment]
I suggest we add a big-data AI to Paperwork, and we crowd-feed it all the dates from all our “words”-files, so that it learns… :-p

**Comment by lapineige** [comment]
I simple comment to say that this kind of feature would be very handy for me: I frequently have to integrate a PDF that is not a fresh scanned document. And sometimes a lot of them at the same time.

Relying on import time is just irrelevant in my case. I can't use it for a search for instance.
So if during import I could activate an option to use the creation/modification time, it would be handy ! :)

I don't understand why you want to scan the date on the document. It would be useful for sure (even if it's more complex) but at least using document date is handy (I mean, the pdf metadata, not the content parsed) as you probably scanned/received it *almost* at the same date as scanned/created.

**Comment by Whisprin** [comment]
Re: 1) Maybe show the dialog on first import and provide a "remember this setting flag" which can be changed in the settings. At least this dialog shouldn't block batch imports of pdfs from folders.

Re: 2) After looking at some recognized text it turns out that a lot of dates are not parsed correctly. e.g.:
`29.o9.2o1e`
There might be an option to pass parameters to the OCR engine to improve the recognition of numbers or suggest similar looking characters. Also a preceding `"Date: "` (in different languages) could be handy.
This already sounds a lot like the suggested big-data AI ;)

**Comment by lapineige** [comment]
> At least this dialog shouldn't block batch imports of pdfs from folders.

:+1:

I think the 1) is a lot easier to implement, while I suppose 2) will require more work/testing.
